Sources/CXUtility/Locking.swift

```diff
@@ -0,0 +1,272 @@
+            return _lock.as(OSUnfairLock.self).tryLock()
+        }
+        #endif
+        return _lock.as(PThreadMutex.self).tryLock()
+    }
+    
+    public func unlock() {
+        #if canImport(Darwin)
+        if #available(macOS 10.12, iOS 10.0, tvOS 10.0, watchOS 3.0, *) {
+            _lock.as(OSUnfairLock.self).unlock()
+            return
+        }
+        #endif
+        _lock.as(PThreadMutex.self).unlock()
+    }
+}
+
+// MARK: - RecursiveLock
+
+public struct RecursiveLock: Locking {
+    
+    private let _lock: UnsafeMutableRawPointer
+    
+    public init() {
+        #if canImport(DarwinPrivate)
+        if #available(macOS 10.14, iOS 12.0, tvOS 12.0, watchOS 5.0, *) {
+            _lock = OSUnfairRecursiveLock().raw
+            return
+        }
+        #endif
+        _lock = PThreadMutex(recursive: true).raw
+    }
+    
+    public func cleanupLock() {
+        #if canImport(DarwinPrivate)
+        if #available(macOS 10.14, iOS 12.0, tvOS 12.0, watchOS 5.0, *) {
+            _lock.as(OSUnfairRecursiveLock.self).cleanupLock()
+            return
+        }
+        #endif
+        _lock.as(PThreadMutex.self).cleanupLock()
+    }
+    
+    public func lock() {
+        #if canImport(DarwinPrivate)
+        if #available(macOS 10.14, iOS 12.0, tvOS 12.0, watchOS 5.0, *) {
+            _lock.as(OSUnfairRecursiveLock.self).lock()
+            return
+        }
+        #endif
+        _lock.as(PThreadMutex.self).lock()
+    }
+    
+    public func tryLock() -> Bool {
+        #if canImport(DarwinPrivate)
+        if #available(macOS 10.14, iOS 12.0, tvOS 12.0, watchOS 5.0, *) {
+            return _lock.as(OSUnfairRecursiveLock.self).tryLock()
+        }
+        #endif
+        return _lock.as(PThreadMutex.self).tryLock()
+    }
+    
+    public func unlock() {
+        #if canImport(DarwinPrivate)
+        if #available(macOS 10.14, iOS 12.0, tvOS 12.0, watchOS 5.0, *) {
+            _lock.as(OSUnfairRecursiveLock.self).unlock()
+            return
+        }
+        #endif
+        _lock.as(PThreadMutex.self).unlock()
+    }
+}
+
+#if canImport(Darwin)
+
+// MARK: - OSUnfairLock
+
+private typealias OSUnfairLock = UnsafeMutablePointer<os_unfair_lock_s>
+
+@available(macOS 10.12, iOS 10.0, tvOS 10.0, watchOS 3.0, *)
+private extension UnsafeMutablePointer where Pointee == os_unfair_lock_s {
+    
+    init() {
+        let l = UnsafeMutablePointer.allocate(capacity: 1)
+        l.initialize(to: os_unfair_lock_s())
+        self = l
+    }
+    
+    func cleanupLock() {
+        deinitialize(count: 1)
+        deallocate()
+    }
+    
+    func lock() {
+        os_unfair_lock_lock(self)
+    }
+    
+    func tryLock() -> Bool {
+        return os_unfair_lock_trylock(self)
+    }
+    
+    func unlock() {
+        os_unfair_lock_unlock(self)
+    }
+}
+
+// MARK: - OSUnfairRecursiveLock
+
+// TODO: Use os_unfair_recursive_lock_s
+#if canImport(DarwinPrivate)
```

Comment:
Doesn't exist for now. We can use `os_unfair_recursive_lock` as soon as it becomes public API.